server-openid4vci/src/main/java/com/android/identity/server/openid4vci/AuthorizeServlet.kt

```diff
@@ -16,56 +40,147 @@ import kotlin.time.Duration.Companion.minutes
-            state.credentialData = CredentialData(
-                firstName = req.getParameter("f"),
-                lastName = req.getParameter("l")
+            val uncompressed = (state.pidReadingKey!!.publicKey as EcPublicKeyDoubleCoordinate).asUncompressedPointEncoding
+            val sessionTranscript = generateBrowserSessionTranscript(
+                Crypto.digest(Algorithm.SHA256, id.toByteArray()),
+                origin,
+                Crypto.digest(Algorithm.SHA256, uncompressed)
             )
+            val deviceResponseRaw = Crypto.hpkeDecrypt(
+                Algorithm.HPKE_BASE_P256_SHA256_AES128GCM,
+                state.pidReadingKey!!,
+                cipherText,
+                sessionTranscript,
+                encapsulatedPublicKey)
+            val parser = DeviceResponseParser(deviceResponseRaw, sessionTranscript)
+            val deviceResponse = parser.parse()
+            val data = NameSpacedData.Builder()
+            for (document in deviceResponse.documents) {
+                for (namespaceName in document.issuerNamespaces) {
+                    for (dataElementName in document.getIssuerEntryNames(namespaceName)) {
+                        val value = document.getIssuerEntryData(namespaceName, dataElementName)
+                        data.putEntry(namespaceName, dataElementName, value)
+                    }
+                }
+            }
+
+            data.putEntry("com.android.identity.server.openid4vci", "extraInfo", Cbor.encode(Tstr(extraInfo)))
+            state.credentialData = data.build()
             storage.update("IssuanceState", "", id, ByteString(state.toCbor()))
         }
         val issuerState = idToCode(OpaqueIdType.ISSUER_STATE, id, 5.minutes)
         resp.sendRedirect("finish_authorization?issuer_state=$issuerState")
     }
+}
+
+// Copied from VerifierServlet.kt as is - should move to some kind of a library
+// TODO: this should be factored into a reusable library.
+
+private const val BROWSER_HANDOVER_V1 = "BrowserHandoverv1"
+private const val ANDROID_CREDENTIAL_DOCUMENT_VERSION = "ANDROID-HPKE-v1"
+
+private fun parseCredentialDocument(encodedCredentialDocument: ByteArray
+): Pair<ByteArray, EcPublicKey> {
+    val map = Cbor.decode(encodedCredentialDocument)
+    val version = map["version"].asTstr
+    if (!version.equals(ANDROID_CREDENTIAL_DOCUMENT_VERSION)) {
+        throw IllegalArgumentException("Unexpected version $version")
+    }
+    val encryptionParameters = map["encryptionParameters"]
+    val pkEm = encryptionParameters["pkEm"].asBstr
+    val encapsulatedPublicKey =
+        EcPublicKeyDoubleCoordinate.fromUncompressedPointEncoding(EcCurve.P256, pkEm)
+    val cipherText = map["cipherText"].asBstr
+    return Pair(cipherText, encapsulatedPublicKey)
+}
+
+//    SessionTranscript = [
+//      null, // DeviceEngagementBytes not available
+//      null, // EReaderKeyBytes not available
+//      AndroidHandover // defined below
```

Comment:
Not defined below... Should this be BrowserHandover?

Reply:
The code is "as is" copy (as stated in a comment) and I am hesitant to change it in any way. This shows very clearly the intent of actually using the same code without any modifications. Once I see differences in cases like this, I always start to suspect that there is some kind of hidden discrepancy.

We should replace this whole block of code anyway.